build/Dockerfile

```diff
@@ -1,4 +1,4 @@
-FROM registry.ci.openshift.org/openshift/release:golang-1.17 as builder
+FROM registry.ci.openshift.org/openshift/release:rhel-8-release-golang-1.19-openshift-4.12 as builder
```

Comment:
This is interesting. I guess there’s no need for this project to back to the bp image; but it wouldn’t hurt anything. What would it take to get this Dockerfile to update automatically with bp?

Reply:
I don't think boilerplate should automatically update this Dockerfile `FROM` statement because the `osd-container-image` convention is meant to be language-agnostic unlike the `golang-osd-operator` convention, which does automatically update the `FROM` statement.

But for background, if I had any choice I'd prefer to `FROM golang:1.19` and output the binary into `scratch` or `distroless`, but in the past I've run into some friction moving away from Red Hat images. To my knowledge `registry.ci.openshift.org/openshift/release:rhel-8-release-golang-1.19-openshift-4.12` is the only pre-baked image with Go 1.19 in it since `registry.ci.openshift.org/openshift/release:golang-1.19` doesn't exist, so that's the only reason I chose it.

Reply:
I'm now remembering that osd-cluster-ready doesn't have an appsre pipeline set up yet. Building and pushing the image, is still a manual process, and deployment is still done via MCC, where "upgrading" entails bumping the image tag therein.

However...

I'm not sure I agree with your assertion that `osd-container-image` is supposed to be language-agnostic. Even if it was, a) we're clearly bound to golang here; and more importantly b) .ci-operator.yaml *is* using the boilerplate backing image. One of the pain points bp is designed to mitigate is the inherent discrepancy between the image built/used for CI and that used for the deployment. It's supposed to keep those as close to synced as possible by editing both .ci-operator.yaml and build/Dockerfile (or whatever your Makefile's `DOCKERFILE` override is set to) to point to the same one. This not only minimizes potential differences between CI and CD; it also obviates the need for a human to remember to cut a followon PR like this one!

Now, I can see that osd-container-image doesn't in fact have the bit that edits the Dockerfile. It should. Do we have a card for that? We should :)

In the meantime I guess this manual process is necessary. But for the reasons above, I still contend that this image should match the one used for CI, which was updated automatically via #29.

Reply:
That's a good point about `.ci-operator.yaml`, I forgot about that...

It was my original intention for `osd-container-image` (that's why it wasn't named `go-container-image` haha) because there are some python containers out there, but you're right that it is pretty much specific to Go.

This has been hooked up though! App-interface will build this container and tag it automatically. We still need to update the new tag manually in MCC for now...that last mile hasn't been automated yet.

Reply:
> because there are some python containers out there

...which could easily use the same backing image for building. And having the backing image synced between CI and CD still has the noted benefits.

> This has been hooked up though!

Ah, nice, I found it in app-interface. So yeah, that uses `make osd-container-image-build-push`, which refers to the `DOCKERFILE` (this file). And we want it to use the same backing image as CI.

...unless you can think of an easy way to override the syncing of .ci-operator.yaml.

But why? The bp image is a superset of the one you want anyway; what harm in just using the bp image for both?

Reply:
> ...which could easily use the same backing image for building. And having the backing image synced between CI and CD still has the noted benefits. (In the case of python, though, I guess you would want to be more careful to make sure the runtime python interpreter was the same for both scenarios.)

Yeah, that's the mess I didn't want to get into - but none of the Python containers are boilerplated anyway, so it's probably a concern that we can ignore.

>...unless you can think of an easy way to override the syncing of .ci-operator.yaml. But why? The bp image is a superset of the one you want anyway; what harm in just using the bp image for both?

I think it's true in this case that the boilerplate image is a supersest of the one I want anyway, but like if Go 1.20 comes out and osd-cluster-ready is ready to upgrade, I wouldn't want to restrict it because boilerplate doesn't support it. On the other hand, maybe boilerplate always should be the first thing updated? The main downside is it's just a fat image and doesn't have an `arm64` version which slows things down for local development, but that applies to `registry.ci.openshift.org/openshift/release:rhel-8-release-golang-1.19-openshift-4.12` as well haha.

Reply:
If go 1.20 comes out
and you update the Dockerfile
but you don't update .ci-operator.yaml
then your CI is lying to your CD.

My primary concern here is to keep those two things in sync.

> it's just a fat image

Okay, this is a valid point... but it would be the impetus for figuring out how to make **CI** run against the smaller image, not justification for using two different images.

> doesn't have an `arm64` version
...
> but that applies to `registry.ci.openshift.org/openshift/release:rhel-8-release-golang-1.19-openshift-4.12` as well haha.

Indeed :)

> maybe boilerplate always should be the first thing updated?

I'll volley back with: boilerplate should be the (only) *mechanism* for updating $[things], where golang version is one of those things.